// brotli-cli/dist/compress.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.compress = void 0;
const utils_js_1 = require("./utils.js");
const anyCompress_js_1 = require("./anyCompress.js");
const compress = async (options) => {
    var _a, _b;
    const buffer = await (0, utils_js_1.readFile)(options.filePath);
    if (buffer.length === 0) {
        return;
    }
    const result = {
        sourceLength: buffer.length,
        compressed: await (0, anyCompress_js_1.anyCompress)(buffer, {
            mode: options.mode,
            quality: options.quality,
            windowSize: options.windowSize,
        }),
    };
    if (result.sourceLength && result.compressed == null) {
        throw new TypeError("Empty response returned from brotli");
    }
    if (options.writeTo === "stdout") {
        process.stdout.write((_a = result.compressed) !== null && _a !== void 0 ? _a : "");
    }
    else {
        await (0, utils_js_1.writeFile)(options.filePath + ".br", (_b = result.compressed) !== null && _b !== void 0 ? _b : "");
    }
};
exports.compress = compress;
